tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp

```diff
@@ -486,38 +486,66 @@ template <typename FnTy>
 void walkRelatedDecls(const ValueDecl *VD, const FnTy &Fn) {
   llvm::SmallDenseMap<DeclName, unsigned, 16> NamesSeen;
   ++NamesSeen[VD->getName()];
-  SmallVector<LookupResultEntry, 8> RelatedDecls;
 
   if (isa<ParamDecl>(VD))
     return; // Parameters don't have interesting related declarations.
 
-  // FIXME: Extract useful related declarations, overloaded functions,
-  // if VD is an initializer, we should extract other initializers etc.
-  // For now we use unqualified lookup to fetch other declarations with the same
-  // base name.
   auto &ctx = VD->getASTContext();
-  auto descriptor = UnqualifiedLookupDescriptor(DeclNameRef(VD->getBaseName()),
-                                                VD->getDeclContext());
-  auto lookup = evaluateOrDefault(ctx.evaluator,
-                                  UnqualifiedLookupRequest{descriptor}, {});
-  for (auto result : lookup) {
-    ValueDecl *RelatedVD = result.getValueDecl();
-    if (RelatedVD->getAttrs().isUnavailable(VD->getASTContext()))
+
+  auto *DC = VD->getDeclContext();
+  if (!DC->isTypeContext()) {
+    SmallVector<LookupResultEntry, 8> RelatedDecls;
+
+    auto descriptor = UnqualifiedLookupDescriptor(DeclNameRef(VD->getBaseName()),
+                                                  DC->getModuleScopeContext());
+
+    auto lookup = evaluateOrDefault(ctx.evaluator,
+                                    UnqualifiedLookupRequest{descriptor}, {});
+
+    for (auto result : lookup) {
+      ValueDecl *RelatedVD = result.getValueDecl();
+      if (RelatedVD->getAttrs().isUnavailable(VD->getASTContext()))
+        continue;
+
+      if (RelatedVD != VD) {
+        ++NamesSeen[RelatedVD->getName()];
+        RelatedDecls.push_back(result);
+      }
+    }
+
+    // Now provide the results along with whether the name is duplicate or not.
+    ValueDecl *OriginalBase = VD->getDeclContext()->getSelfNominalTypeDecl();
+    for (auto Related : RelatedDecls) {
+      ValueDecl *RelatedVD = Related.getValueDecl();
+      bool SameBase = Related.getBaseDecl() && Related.getBaseDecl() == OriginalBase;
+      Fn(RelatedVD, SameBase, NamesSeen[RelatedVD->getName()] > 1);
+    }
+
+    return;
+  }
+
+  SmallVector<ValueDecl *, 4> results;
+
+  auto type = DC->getDeclaredInterfaceType();
+  if (!type->is<ErrorType>()) {
+    DC->lookupQualified(type, DeclNameRef(VD->getBaseName()),
+                        NL_QualifiedDefault, results);
+  }
+
+  SmallVector<ValueDecl *, 8> RelatedDecls;
+  for (auto result : results) {
+    if (result->getAttrs().isUnavailable(ctx))
```

Comment:
This whole function could be easier to understand with either a comment describing the overall strategy, or by breaking it up.

Reply:
You're right. I completely refactored this function, will push a new version shortly.

Reply:
Thanks!